docs/site/tutorials/core/10-advanced-recipes.md

```diff
@@ -47,3 +61,68 @@ const interceptor: Interceptor = (invocationCtx, next) => {
 ```
 
 # Magic ValueOrPromise
+
+For the dependency injection framework, there are two flavors: synchronous and
+asynchronous.
+
+`Context.getSync()` resolves a value synchronously, whereas `Context.get()`
+allows you to resolve the value asynchronous and returns a `Promise`.
+
+When `ValueOrPromise` is being used in a value provider, if the value is
+produced synchronously, a value will be returned, otherwise a Promise will be
+return. When you are chaining multiple value and if any of them is a Promise,
+you'll be getting a Promise.
+
+Using the `ChineseGreeter` as an example:
+
+```ts
+/**
+ * A greeter implementation for Chinese.
+ */
+@bind(asGreeter)
+export class ChineseGreeter implements Greeter {
+  language = 'zh';
+
+  constructor(
+    /**
+     * Inject the configuration for ChineseGreeter
+     */
+    @config()
+    private options: ChineseGreeterOptions = {nameFirst: true},
+  ) {}
+
+  greet(name: string) {
+    if (this.options && this.options.nameFirst === false) {
+      return `你好，${name}！`;
+    }
+    return `${name}，你好！`;
+  }
+}
+```
+
+There are two ways to configure the greeter.
+
+**Option 1**
+
+We call `app.getSync('greeters.ChineseGreeter')` or
+`app.get('greeters.ChineseGreeter')` to get the `ChineseGreeter`.
+
+```ts
+app.configure('greeters.ChineseGreeter').to({nameFirst: false});
+```
+
+**Option 2**
+
+We can only call `app.get('greeters.ChineseGreeter')` because the configuration
+dependencies is asynchronous.
```

Comment:
Please note `app.getSync` will throw an exception to indicate that the ChineseGreeter binding cannot be resolved synchronously.